File: vectorcloud/plugins/stats.py
import anki_vector
from flask_socketio import emit
from vectorcloud import socketio
from vectorcloud.main.models import Vectors
from vectorcloud.main.utils import run_plugin


class Plugin:
    def __init__(self, *args, **kwargs):
        # parse user supplied plugin settings
        for key, value in kwargs.items():
            self.__dict__[key] = value

        # set defaults for omitted options
        if not hasattr(self, "vector_id"):
            self.vector_id = "all"
        if not hasattr(self, "log"):
            self.log = "true"

    def interface_data(self):
        interface_data = {
            "plugin_settings": [
                {
                    "name": "vector_id",
                    "default": "all",
                    "description": "which vector id to use for the command",
                },
                {
                    "name": "log",
                    "default": "true",
                    "description": "create a log item when plugin is ran",
                },
            ],
            "plugin_description": "None",
            "plugin_icons": [
                {
                    "mdi_class": "refresh",
                    "class": "stats-refresh-btn",
                    "tooltip": "Refesh stats",
                },
                {"mdi_class": "info", "class": "stats-info-btn", "tooltip": "Stats"},
                {"mdi_class": "pages", "class": "stats-cube-btn", "tooltip": "Cube"},
                {
                    "mdi_class": "battery_full",
                    "class": "stats-battery-btn color-ignore",
                    "tooltip": "Battery",
                },
            ],
            "plugin_panels": [
                {"class": "stats-info-panel", "template": "stats-info-panel.html"},
                {"class": "stats-cube-panel", "template": "stats-cube-panel.html"},
                {
                    "class": "stats-battery-panel",
                    "template": "stats-battery-panel.html",
                },
            ],
            "plugin_js": ["stats.js"],
            "plugin_dependencies": ["logbook", "cube"],
        }
        return interface_data

    def on_startup(self):
        @socketio.on("request_stats")
        def handle_stats_request(json):
            if "vector_id" not in json:
                json["vector_id"] = "all"

            run_plugin("stats", {"vector_id": json["vector_id"]})

    def run(self):
        if self.vector_id == "all":
            vectors = Vectors.query.all()
        else:
            vectors = [Vectors.query.filter_by(id=self.vector_id).first()]

        responses = []
        for vector in vectors:
            response = {
                "id": vector.id,
                "ip": vector.ip,
                "name": vector.name,
                "serial": vector.serial,
            }
            robot = anki_vector.Robot(vector.serial, behavior_control_level=None)
            try:
                robot.connect()
            except Exception as e:
                run_plugin(
                    "logbook",
                    {
                        "name": f"{vector.name} failed to connect",
                        "vector_id": vector.id,
                        "info": str(e),
                        "log_type": "fail",
                    },
                )
                continue

            version_state = robot.get_version_state()
            battery_state = robot.get_battery_state()

            response["version"] = version_state.os_version
            response["battery_voltage"] = battery_state.battery_volts
            response["battery_level"] = battery_state.battery_level
            response["status_charging"] = battery_state.is_on_charger_platform
            response["cube_battery_level"] = battery_state.cube_battery.level
            response["cube_id"] = battery_state.cube_battery.factory_id
            response["cube_battery_volts"] = battery_state.cube_battery.battery_volts

            emit("stats", response, broadcast=True, namespace="/")
            if self.log == "true":
                run_plugin(
                    "logbook",
                    {
                        "name": f"{vector.name} reported its stats",
                        "vector_id": vector.id,
                        "info": str(response),
                        "log_type": "success",
                    },
                )
            responses.append(response)

        if len(responses) > 1:
            return responses
        else:
            return response

File: vectorcloud/static/js/plugins/stats.js
socket.on('connect', function() {
    socket.emit('request_stats', {vector_id: "all"});
});

socket.on('stats', function(stats) {
    $(".plugin-panels-container").each(function(e) {
       if ($(this).attr('vector_id') == stats.id) {
           $("#status-connected-" + stats.name).attr("src", 'static/images/icons/vector-green.svg')
           $("#status-name-" + stats.name).text(stats.name);

           $(this).find('.status-info').empty();
           $(this).find('.status-info').append(
               `
                IP: ${stats.ip} <br>
                OS Version: ${stats.ip} <br>
                Serial: ${stats.serial} <br>
               `
           );
           $(this).find('.status-cube').empty();
           $(this).find('.status-cube').append(
               `
                Cube ID: ${stats.cube_id} <br>
                Cube Battery Volts: ${stats.cube_battery_volts} <br>
                Cube Battery Level: ${stats.cube_battery_level} <br>
               `
           );
           $(this).find('.status-battery').empty();
           $(this).find('.status-battery').append(
               `
                Battery Voltage: ${stats.battery_voltage} <br>
                Battery Level: ${stats.battery_level} <br>
                Charging Status: ${stats.status_charging} <br>
               `
           );
       }
    });
    $(".plugin-icons-container").each(function(e) {
        if ($(this).attr('vector_id') == stats.id) {
            if (stats.status_charging == true){
                $(this).find('.stats-battery-btn').removeClass('theme-secondary-text');
                $(this).find('.stats-battery-btn').addClass('theme-primary-text');
                $(this).find('.stats-battery-btn').text('battery_charging_full');
            } else if (stats.battery_level < 2){
                $(this).find('.stats-battery-btn').addClass('theme-secondary-text');
                $(this).find('.stats-battery-btn').removeClass('theme-primary-text');
                $(this).find('.stats-battery-btn').text('battery_alert');
            } else if (stats.battery_level >= 2){
                $(this).find('.stats-battery-btn').addClass('theme-secondary-text');
                $(this).find('.stats-battery-btn').removeClass('theme-primary-text');
                $(this).find('.stats-battery-btn').text('battery_full');
            }
        }
    });
});

$( document ).ready(function() {
    setInterval(function(){
        socket.emit('request_stats', {vector_id: "all"});
    }, 60000);

    $(".stats-info-btn").on('click', function(e) {
        var vector_id = $(this).closest('.plugin-icons-container').attr('vector_id');
        show_panel(".stats-info-panel", vector_id, $(this))
    });
    $(".stats-cube-btn").on('click', function(e) {
        var vector_id = $(this).closest('.plugin-icons-container').attr('vector_id');
        show_panel(".stats-cube-panel", vector_id, $(this))
    });
    $(".stats-battery-btn").on('click', function(e) {
        var vector_id = $(this).closest('.plugin-icons-container').attr('vector_id');
        show_panel(".stats-battery-panel", vector_id)
    });
    $(".connect-to-cube-btn").on('click', function(e) {
        var vector_id = $(this).closest('.plugin-panels-container').attr('vector_id');
        socket.emit('run_plugin', {
            name: "cube",
            command: 'connect',
            vector_id: vector_id
        });
    });
    $(".stats-refresh-btn").on('click', function(e) {
        var vector_id = $(this).closest('.plugin-icons-container').attr('vector_id');
        socket.emit('run_plugin', {
            name: "stats",
            vector_id: vector_id
        });
    });
});